README.txt
This product makes is easy to send files by e-mail on top of a CPS portal

Dependencies: CPSSubscriptions

HOW TO INSTALL
--------------

Import CPSFileMailer 'default' profile and you're set

Optionaly, you can add the CPSFileMailer 'bigfile' profile, if you're using
this in conjunction with CPSTramline 'bigfile'.
See the end of this document for details on the latter.

OPERATION
---------
It replaces the standard "notify_by_email" action so that
the created link serves the file directly with a token argument,
and no actual login. 

Since there's no other security than the token, you have to teach your users 
the obvious *confidentiality* issues.

The token lifetime can be specified in the portal_filemailer tool properties.
There's also an option to use a different base url in the link than the one used by
the current user.

The link provides the same revision that the user would see by navigation to the
document with his language settings

BEHIND THE SCENES
-----------------

A directory (filetokens) is used to keep the relation between 
doc id, token, expiration date. 

The default profiles installs a ZODB directory, but you're free to replace it 
with another kind (provided that DateTime objects are correctly supported)

You should regularly purge outdated tokens, which can be done by a simple GET on 
<portal_url>/portal_filemailer/purgeOutdatedTokens (Manager auth required). Don't 
forget to add ?__disable_cookie_login=1 if you want to perform a BASIC 
authentication (convenient with tools such as wget). 


THE BIGFILE PROFILE
-------------------
Instead of replacing the standard 'notify_by_email', this profile adds a new action
'send_by_email' (so that File documents have both) and is ready to serve documents
of type 'Big File', as provided by CPSTramline's 'bigfile' profile.




CHANGES
Requires
~~~~~~~~
- Zope 2.10 / CMF 2.1.1
New features
~~~~~~~~~~~~
-
Bug fixes
~~~~~~~~~
- Added a test for skins layer ordering
New internal features
~~~~~~~~~~~~~~~~~~~~~
- 


tests/layer.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: __init__.py 890 2008-06-18 18:26:32Z joe $

from Testing import ZopeTestCase

from Products.CPSDefault.tests.CPSTestCase import (
    CPSTestCase,
    ExtensionProfileLayerClass)

# register profiles
ZopeTestCase.installProduct('CPSFileMailer')
ZopeTestCase.installProduct('CPSSubscriptions')

class CPSFileMailerLayerClass(ExtensionProfileLayerClass):
    extension_ids = ('CPSSubscriptions:default', 'CPSFileMailer:default', )

CPSFileMailerLayer = CPSFileMailerLayerClass(
    __name__,
    'CPSFileMailerLayer'
    )


tests/test_tool.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: __init__.py 890 2008-06-18 18:26:32Z joe $

import unittest
from Products.CMFCore.utils import getToolByName
from Products.CPSDefault.tests.CPSTestCase import CPSTestCase
from Products.CPSFileMailer.tests.layer import CPSFileMailerLayer

from OFS.Image import File
from AccessControl import Unauthorized
from Products.CPSFileMailer.tool import FileMailerTool

class FileMailerToolTestCase(unittest.TestCase):
    def setUp(self):
        self.fmtool = FileMailerTool()
        
    def test_fields_mapping(self):
        self.fmtool.manage_changeProperties(
            fields_mapping=('Type:attached', 'Other:file '))
        self.assertEquals(self.fmtool.fields_mapping_c, {'Type': 'attached', 
                                                         'Other': 'file'})
        self.fmtool.manage_changeProperties(
            fields_mapping=('', 'Type:attached', ' '))
        self.assertEquals(self.fmtool.fields_mapping_c, {'Type': 'attached'}) 

        self.fmtool.manage_changeProperties(fields_mapping=())
        self.assertEquals(self.fmtool.fields_mapping_c, {}) 

class FileMailerToolIntegrationTestCase(CPSTestCase):

    layer = CPSFileMailerLayer

    def afterSetUp(self):
        self.login('manager')
        self.ws = self.portal.workspaces
        self.wftool = self.portal.portal_workflow
        self.fmtool = self.portal.portal_filemailer
        f = File('file', 'truc.txt', 'here it is')
        docname = self.wftool.invokeFactoryFor(self.ws, 'File', 'item',
                                                file=f)
        self.doc = getattr(self.ws, docname)

    def testSkins(self):
        # check that our skin layer is before cpsdefault
        # actually this is a test of CPSCore's adapter for the skins tool
        # and should be done cleanly in CPSCore tests (no time right now, TODO)
        sktool = getToolByName(self.portal, 'portal_skins')
        path = sktool._getSelections()['CPSSkins'].split(',')
        self.failIf(path.index('cps_filemailer') > path.index('cps_default'))

    def testCreateToken(self):
        token, _, _ = self.fmtool.createTokenFor(self.doc)
        tdir = self.portal.portal_directories.filetokens
        entry = tdir._getEntry(token)
        self.assertEquals(entry.get('doc'), self.doc.getContent().getId())
        self.assertTrue(entry.get('expiration').isFuture())

    def testRetrieveFile(self):
        token, _, _ = self.fmtool.createTokenFor(self.doc)
        f = self.fmtool.retrieveFile(token)
        self.assertFalse(f is None)
        self.assertEquals(f.title, 'truc.txt')

    def testRetrieveOtherFile(self):
        self.fmtool.manage_changeProperties(fields_mapping=('File:other',))
        self.doc.getContent().other = File('other', 'autre.pdf', 'content')
        token, _, _ = self.fmtool.createTokenFor(self.doc)
        f = self.fmtool.retrieveFile(token)
        self.assertFalse(f is None)
        self.assertEquals(f.title, 'autre.pdf')
        self.assertEquals(str(f), 'content')

    def testTraversal(self):
        token, filename, _ = self.fmtool.createTokenFor(self.doc)
        self.logout()

        # now modelling what happens if client follows link
        # first traverse, finally call index_html
        base = self.portal.absolute_url()
        self.portal.REQUEST.form['token'] = token

        path = self.portal.portal_filemailer.getPhysicalPath() + (filename,)
        # BaseRequest.traverse() needs a non empty list of parents, whose
        # last item is more or less the beginning of traversal
        self.portal.REQUEST['PARENTS'] = [self.app]
        meth = self.portal.REQUEST.traverse('/'.join(path))
        self.assertEquals('here it is', meth())

        # Now, a restrictedTraverse from code (yes, that's a totally different
        # piece of code being executed)
        fs = self.portal.portal_filemailer.restrictedTraverse(filename)
        self.assertEquals('here it is', fs.index_html())

    def testWrongToken(self):
        token, _, _ = self.fmtool.createTokenFor(self.doc)
        self.assertRaises(Unauthorized, self.fmtool.retrieveFile, '000')

    def testExpired(self):
        # trick the took into setting the expiration in the past
        self.fmtool.tok_lifetime = -1
        token, _, exp = self.fmtool.createTokenFor(self.doc)
        self.assertTrue(exp.isPast())
        self.assertRaises(Unauthorized, self.fmtool.retrieveFile, token)

    def testPurge(self):
        tdir = self.portal.portal_directories.filetokens
        # two valid tokens
        self.fmtool.createTokenFor(self.doc)
        self.fmtool.createTokenFor(self.doc)
        self.fmtool.tok_lifetime = -1
        # one outdated one
        self.fmtool.createTokenFor(self.doc)

        # reality check
        self.assertEquals(3, len(tdir.listEntryIds()))
        removed, kept = self.fmtool._purgeOutdatedTokens()
        self.assertEquals(removed, 1) 
        self.assertEquals(kept, 2) 
        # reality check
        self.assertEquals(2, len(tdir.listEntryIds()))

    def testPurgeTTW(self):
        # one valid token
        self.fmtool.createTokenFor(self.doc)
        self.fmtool.tok_lifetime = -1
        # two outdated ones
        self.fmtool.createTokenFor(self.doc)
        self.fmtool.createTokenFor(self.doc)
        line = self.fmtool.purgeOutdatedTokens()
        self.assertTrue(isinstance(line, str))

def test_suite():
    return unittest.TestSuite((
        unittest.makeSuite(FileMailerToolTestCase),
        unittest.makeSuite(FileMailerToolIntegrationTestCase),
        ))


__init__.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: __init__.py 890 2008-06-18 18:26:32Z joe $

from AccessControl import ModuleSecurityInfo

from Products.CMFCore import utils
from Products.CMFCore.DirectoryView import registerDirectory

from Products.GenericSetup import profile_registry
from Products.GenericSetup import EXTENSION

from Products.CPSCore.interfaces import ICPSSite
from Products.CPSCore.upgrade import registerUpgradeCategory

registerDirectory('skins', globals())

def initialize(registrar):
    """Initialize Paris Montagne Contacts content and tools.
    """

    profile_registry.registerProfile(
        'default',
        'CPS File Mailer',
        "CPS File Mailer, default configuration",
        'profiles/default',
        'CPSFileMailer',
        EXTENSION,
        for_=ICPSSite)

    profile_registry.registerProfile(
        'bigfile',
        'CPS File Mailer Big File',
        "CPS File Mailer for CPSTramline's Big File",
        'profiles/bigfile',
        'CPSFileMailer',
        EXTENSION,
        for_=ICPSSite)


interfaces.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: __init__.py 890 2008-06-18 18:26:32Z joe $

from zope.interface import Interface

class IFileMailerTool(Interface):
    """FileMailer Tool.
    """

    def createTokenFor(proxy):
        """Create a new token entry in the filetokens directory.
        
        Return token and file name
        """

    def retrieveFile(token):
        """Retrieve the OFS.Image.File object associated with the token."""

    def getBaseUrl():
        """Return base url prop or standard portal base url."""

    def purgeOutdatedTokens(REQUEST=None):
        """Purge outdated tokens and return a simple status line."""

    def _purgeOutdatedTokens():
        """Purge outdated tokens and return (number of removed, of kept)"""


exportimport.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: __init__.py 890 2008-06-18 18:26:32Z joe $

# Zope3 component architecture
from zope.component import adapts
from zope.interface import implements

# Standard GenericSetup base classes and functions
from Products.GenericSetup.utils import exportObjects
from Products.GenericSetup.utils import importObjects
from Products.GenericSetup.utils import XMLAdapterBase
from Products.GenericSetup.utils import PropertyManagerHelpers

from Products.CMFCore.utils import getToolByName

from Products.CPSUtil.PropertiesPostProcessor import (
    PostProcessingPropertyManagerHelpers)

# Genericsetup multi adapts a class that implement IVogonTool and a particular
# ISetupEnvironment to and IBody (piece of XML configuration).
from Products.GenericSetup.interfaces import IBody
from Products.GenericSetup.interfaces import ISetupEnviron
from Products.CPSFileMailer.interfaces import IFileMailerTool

TOOL = 'portal_filemailer'
NAME = 'filemailer'

def exportFileMailerTool(context):
    """Export our Filemailer tool configuration
    """
    site = context.getSite()
    tool = getToolByName(site, TOOL, None)
    if tool is None:
        logger = context.getLogger(NAME)
        logger.info("Nothing to export.")
        return
    exportObjects(tool, '', context)

def importFileMailerTool(context):
    """Import filemailer tool configuration
    """
    site = context.getSite()
    tool = getToolByName(site, TOOL)
    importObjects(tool, '', context)


# This the XMLAdapter itself. It encodes the im- / export logic that is specific
# to our tool. `im- / exportObjects` functions will find it thanks to the zope
# components machinery and the associations made in the configure.zcml file.

class FileMailerToolXMLAdapter(XMLAdapterBase, 
                               PostProcessingPropertyManagerHelpers):
    """XML importer and exporter for the Filemailer tool.

    The filemailer tool has very little persistent configuration to im- / export
    in itself. This mostly about sub objects.
    """

    adapts(IFileMailerTool, ISetupEnviron)
    implements(IBody)

    _LOGGER_ID = NAME
    name = NAME

    def _exportNode(self):
        """Export the object as a DOM node.
        """
        node = self._getObjectNode('object')
        node.appendChild(self._extractProperties())
        self._logger.info("Filemailer tool exported.")
        return node

    def _importNode(self, node):
        """Import the object from the DOM node.
        """
        if self.environ.shouldPurge():
            self._purgeProperties()
        self._initProperties(node)
        
        self._logger.info("Filemailer tool imported.")


skins/cfm_sendmail.py
##parameters=REQUEST=None, **kw
# $Id: content_notify_sendmail.py 50671 2006-12-21 13:06:04Z div $
"""Send a email for custom notifications

kws should contain:
members, explicit_recipients_emails, mail_subject, mail_body
"""

from Products.CMFCore.utils import getToolByName
from Products.CPSUtil.text import get_final_encoding

if REQUEST is not None:
    kw.update(REQUEST.form)

members = kw.get('members', [])
explicit_recipients_emails = kw.get('explicit_recipients_emails', [])
mail_subject = kw.get('mail_subject', '')
mail_body = kw.get('mail_body', '')

fmtool = context.portal_filemailer
cpsmcat = context.translation_service
encoding = get_final_encoding(cpsmcat)

token, filename, exp = fmtool.createTokenFor(context) 
base_url = fmtool.getBaseUrl()
mail_body = mail_body + '\n%s/cfm_get?token=%s&file=%s\n%s\n' % (
    base_url, token, filename, exp.strftime(
        cpsmcat('cpsfilemailer_token_expiration_msg').encode(encoding)))

mtool = getToolByName(context, 'portal_membership')
member = mtool.getAuthenticatedMember()
reply_to_email = mtool.getEmailFromUsername(member.getUserName())
reply_to_name = member.getProperty('sn') + ' ' + member.getProperty('givenName')

subtool = getToolByName(context, 'portal_subscriptions')
sender_email, sender_name = subtool.getMailSenderInfo()

tos = explicit_recipients_emails or []
for member_id in members:
    email = mtool.getEmailFromUsername(member_id)
    if email is not None:
        tos.append(email)

to_str = ','.join(tos)

infos = {
    'sender_email'   : sender_email,
    'sender_name'    : sender_name,
    'reply_to_email' : reply_to_email,
    'reply_to_name'  : reply_to_name,
    'body'           : (mail_body, 'text/plain'),
    'subject'        : mail_subject,
    'to'             : ','.join(tos),
    }

cerror = subtool.sendmail(infos=infos)


if REQUEST is not None:
    psm = 'psm_an_email_has_been_sent'
    redirect_url = REQUEST['URL1'] + '?portal_status_message=%s' % psm
    return REQUEST.RESPONSE.redirect(redirect_url)



skins/cfm_email_do_search_results.py
##parameters=dir, datastructure, **kw

datamodel = datastructure.getDataModel()

original_mapping = {}
for key, value in datamodel.items():
    if value:
        original_mapping[key] = value

result_fields = context.getDirectoryResultFields(dir.getId(), dir.title_field)

aclu = context.acl_users
return_fields = []
sort_by = None
sort_direction = None
process_fields = {}
for field in result_fields:
    return_fields.append(field['id'])
    sorted = field.get('sort')
    if sorted == 'asc':
        sort_by = field['id']
        sort_direction = 'asc'
    elif sorted == 'desc':
        sort_by = field['id']
        sort_direction = 'desc'
    if field.get('process'):
        process_fields[field['id']] = field['process']

call_context = kw.get('call_context')
mtool = context.portal_membership

users_mapping = original_mapping.copy()

if call_context is not None:
    dict_roles = mtool.getMergedLocalRoles(call_context, withgroups=0)
    search_restricted_list = dict_roles.keys()
    users_mapping[dir.id_field] = search_restricted_list
results = dir.searchEntries(return_fields=return_fields, **users_mapping)

#
# Check if there's groups with local roles
#

dict_roles = mtool.getMergedLocalRoles(call_context, withgroups=1)
groups = [x[len('group:'):] for x in dict_roles if x.startswith('group:')]

# special handling of special groups to behave as the RecipientRules logic
if "role:Authenticated" in groups and not aclu.is_role_authenticated_empty:
    results = dir.searchEntries(return_fields=return_fields, **original_mapping)
    skip_groups = True
elif "role:Anonymous" in groups and not aclu.is_role_anonymous_empty:
    results = dir.searchEntries(return_fields=return_fields, **original_mapping)
    skip_groups = True
else:
    skip_groups = False

if not skip_groups:
    for group_id  in groups:
        group_mapping = original_mapping.copy()
        group_mapping['groups'] = [group_id]
        group_member_results = dir.searchEntries(return_fields=return_fields,
                                                 **group_mapping)
        for member in group_member_results:
            if member not in results:
                results.append(member)

for field, process_meth in process_fields.items():
    meth = getattr(context, process_meth, None)
    if not meth:
        continue
    for item in results:
        value = item[1].get(field)
        item[1][field] = meth(field, value)

if sort_by:
    items = [(item[1].get(sort_by), item) for item in results]
    items.sort()
    if sort_direction == 'desc':
        items.reverse()
    results = [item[1] for item in items]

rendered = dir.cfm_email_search_results(results=results)

return rendered, 'results'


tool.py
# (C) Copyright 2008 Tonal
# Author: Georges Racinet <georges@racinet.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: tool.py 973 2008-10-20 07:03:15Z joe $

import os
import errno
import sys
import random

import Acquisition
from AccessControl import ClassSecurityInfo
from AccessControl import Unauthorized
from Globals import InitializeClass
from DateTime.DateTime import DateTime
from zExceptions import BadRequest

from zope.interface import implements

from Products.CMFCore.utils import SimpleItemWithProperties
from Products.CMFCore.utils import getToolByName
from Products.CMFCore.utils import UniqueObject
from Products.CMFCore.utils import _checkPermission
from Products.CMFCore.permissions import View
from Products.CMFCore.permissions import ManagePortal

from Products.CPSUtil.PropertiesPostProcessor import PropertiesPostProcessor

from interfaces import IFileMailerTool

TOKEN_DIR = 'filetokens'

class FileMailerTool(UniqueObject, PropertiesPostProcessor, 
                     SimpleItemWithProperties):
    id = 'portal_filemailer'
    meta_type = 'File Mailer Tool'

    implements(IFileMailerTool)

    security = ClassSecurityInfo()

    manage_options = SimpleItemWithProperties.manage_options + (
        {'label': 'Export',
         'action': 'manage_genericSetupExport.html',
         },
        )

    _propertiesBaseClass = SimpleItemWithProperties
    _properties=(
            {'id': 'tok_lifetime', 'type': 'int', 'mode':'w',
             'label': 'Lifetime of tokens (hours)'},
            {'id': 'base_url', 'type': 'string', 'mode':'w',
             'label': 'Public portal url (for emails, in case it differs from what contributors see)'},
            {'id': 'fields_mapping', 'type': 'lines', 'mode':'w',
             'label': 'Fields to retrieve (lines of <portal_type>:<field id>)'},
            )
    tok_lifetime = 12
    base_url = ''
    fields_mapping = ()
    fields_mapping_c = {}

    def _postProcessProperties(self):
        """Extract the fields_mapping dict."""

        self.fields_mapping = tuple(l for l in self.fields_mapping if l.strip())
        self.fields_mapping_c = dict(l.strip().split(':') 
                                     for l in self.fields_mapping)

    def _getTokensDirectory(self):
        return getToolByName(self, 'portal_directories')._getOb(TOKEN_DIR)

    security.declarePublic('getBaseUrl')
    def getBaseUrl(self):
        """Return base url. 

        Defaults to standard meaning of 'base_url."""
        return self.getProperty('base_url') or getToolByName(
            self, 'portal_url').getPortalObject().absolute_url()

    security.declarePublic('createTokenFor')
    def createTokenFor(self, proxy):
        """See interface."""

        # GR better to have the calling script issue the getContent()
        # and hence trigger security check ?
        if not _checkPermission(View, proxy):
            raise Unauthorized
        doc = proxy.getContent()
        docid = doc.getId() # includes rev
        exp = DateTime() + self.tok_lifetime / 24.0
        entry = { 'doc': docid,
                  'expiration': exp}

        tdir = self._getTokensDirectory()
        # now creation. In theory, the try/except approach should be faster
        # than using _hasEntry (avoid a stupid db miss). In practice,  most
        # dir implementations of _createEntry do call _hasEntry first (at least
        # we got it just once
        while True:
            token =  ''.join((hex(random.randrange(sys.maxint))[2:]
                                  for i in range(3)))
            entry['token'] = token
            try:
                tdir._createEntry(entry)
            except KeyError: 
                pass
            else:
                break

        return token, self._extractFile(doc).title, exp
        
    def _purgeOutdatedTokens(self):
        """See interface.

        Dumb implementation based on the poor ZODBDirectory request 
        capabilities."""

        tdir = self._getTokensDirectory()
        removed, kept = 0, 0
        # See #1927: searchEntries and return fields
        for tok in tdir.listEntryIds():
            entry = tdir._getEntry(tok)
            if entry['expiration'].isPast():
                removed += 1
                tdir._deleteEntry(tok)
            else:
                kept += 1
        return removed, kept

    security.declareProtected(ManagePortal, 'purgeOutdatedTokens')
    def purgeOutdatedTokens(self, REQUEST=None):
        """See interface."""

        if REQUEST is not None:
            REQUEST.RESPONSE.setHeader('Content-Type', 'text/plain')
        return "Removed %d outdated tokens and kept %d active ones.\n" % self._purgeOutdatedTokens()

    security.declarePublic('retrieveDoc')
    def retrieveDoc(self, token):
        """See interface.
        """
        tdir = self._getTokensDirectory()
        try:
            entry = tdir._getEntry(token)
        except KeyError: 
            raise Unauthorized
        
        if entry['expiration'].isPast():
            raise Unauthorized
        
        repotool = getToolByName(self, 'portal_repository')
        try:
            doc = repotool._getOb(entry['doc'])
        except KeyError, AttributeError:
            raise Unauthorized

        return doc

    security.declarePublic('retrieveFile')
    def retrieveFile(self, token):
        """See interface.
        """
        return self._extractFile(self.retrieveDoc(token))

    def _extractFile(self, doc):
        """Return the File object for doc according to conf."""
        ptype = doc.portal_type
        field = self.fields_mapping_c.get(doc.portal_type)
        if field is None:
            raise ValueError("portal_type '%s' unknown" % ptype) 
        
        # bypass of DataModel to avoid security problems
        return getattr(doc, field, None)

    def __getitem__(self, name):
        """Retrieve file if needed and wrap it in a FileServer instance.
        """
        req = self.REQUEST
        token = req.form.get('token')
        if token is None:
            return super().__getitem__(name)

        f = self.retrieveFile(token)
        if f.title_or_id() != name:
            # I can't see any legit use of this, but a few nasty ones
            raise BadRequest(
                "! File name changed. Should have been '%s'" % f.title_or_id())
        return FileServer(f, req).__of__(self)


class FileServer(Acquisition.Explicit):

    security = ClassSecurityInfo()

    def __init__(self, f, req):
        self.file = f
        self.req = req

    security.declarePublic('index_html')
    def index_html(self):
        """Serve the file."""
        resp = self.req.RESPONSE
        resp.setHeader('Content-Disposition', 
                       'attachment; filename=%s' % self.file.title_or_id())
        return self.file.index_html(self.req, resp)


        
